### PR TITLE
Make the `showPages` prop on our Pagination components responsive

### PR DESCRIPTION
**Pull request by mperrotti**
The `showPages` prop on both Pagination components can now accept a responsive value.

Closes https://github.com/primer/react/issues/2773

### Screenshots

https://github.com/primer/react/assets/2313998/1942a266-0c39-45d3-a60f-8ef07eaa8f53

### Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation
- [x] Changes are [SSR compatible](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#ssr-compatibility)
- [x] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

**Comment by mperrotti** [comment]
@joshblack - instead of building it in, I wanted to give users the option to change the value for different viewport sizes. For example, I might be using pagination in a narrow sidebar on wide screens, so I'd want to set `showPages={{wide: false}}`

**Comment by joshblack** [comment]
@mperrotti would this be a good case for container queries where we could make that decision based on the container size? The ideal here seems to be that we should show the pages if there is space and hide them if there is not and that could be a way of baking that in without having to customize it / tie it to media queries if the width isn't really tied to the viewport width.

**Comment by mperrotti** [comment]
@joshblack - I agree that this could be a great use for container queries. Do you think I should refactor before we merge this?

**Comment by joshblack** [comment]
@mperrotti if we end up using container queries, would it be that `showPages` would still be a `prop` or would this be something that is handled automatically? (only show pages if there is space)

I think if, with container queries, this behavior becomes automatic that'd be a huge win but if it's just a different way to implement `showPages` then I totally would get thinking of this as an implementation detail.

**Comment by mperrotti** [comment]
> I think if, with container queries, this behavior becomes automatic that'd be a huge win

Great idea! I think it makes sense to just make this behavior automatic.

**Comment by mperrotti** [comment]
@joshblack - `styled-components` won't support container queries until v6 🤬

https://github.com/styled-components/styled-components/issues/3911

Should we:
1. hold off on this until we upgrade to styled-components v6
2. merge as-is and make a breaking API change later if we still think it's better to automatically hide the page numbers for containers of a certain size
3. merge as-is and plan to keep this new `showPages` API
4. something else

There are trade-offs to all of those options, but I think **option 2** is looking the best to me right now. If you have any other ideas about how we might work around this, let me know!

File: src/DataTable/Pagination.tsx
Comment (joshblack) [inline src/DataTable/Pagination.tsx:241]: Is the `hiddenViewportRanges` prop something that can be placed on `TablePaginationSteps`? Would be great to calculate this once and place it on the outermost node and then target it e.g.

```css
  @media ${viewportRanges.narrow} {
    .TablePaginationSteps[data-hidden-viewport-ranges*='narrow'] > *:not(:first-child):not(:last-child) {
      display: none;
    }

    .TablePaginationSteps[data-hidden-viewport-ranges*='narrow'] > *:first-child {
      margin-inline-end: 0;
    }
  }
```

If not, I think this is something we should calculate once instead of as a function and just pass its value to each `Step` instead of messing around with `useCallback()` and calling it for each step.

**Review by joshblack — COMMENTED** [review]
With the stuff from container queries figured moving forward with option 2 works like you suggested would work! Just had one question on the implementation, curious what you think 👀

File: src/DataTable/Pagination.tsx
Comment (mperrotti) [reply src/DataTable/Pagination.tsx:241]: Ahh good catch! Ok, I'll refactor 👍

**Review by langermank — APPROVED** [review]

**Review by joshblack — APPROVED** [review]
🥳

**Pull request merged**
